### ISSUE TITLE
Link error when adding SFML as a subdirectory and with a static runtime

### ISSUE THREAD

**Issue by baugid** [SFML/SFML#1679]
## Subject of the issue

When adding SFML as a subdirectory to a cmake project and compiling it with MSVC using the static runtime a link time error occurs (`LNK2038 mismatch detected for 'RuntimeLibrary': value 'MD_DynamicRelease' doesn't match value 'MT_StaticRelease' in main.obj`)

## Environment

* Windows 10 64bit version 1909
* current SFML git master
* current version of MSVC/Visual studio 2019
* Tried to use /MT compiler flag

## Steps to reproduce
Just try to build the following code with MSVC and clone SFML to a directory called SFML

main.cpp:
```cpp
#include <SFML/Graphics.hpp>

int main()
{
    sf::RenderWindow window(sf::VideoMode(1280, 720), "Example code");
    window.setFramerateLimit(60);

    while (window.isOpen())
    {
        sf::Event event;
        while (window.pollEvent(event))
        {
            if (event.type == sf::Event::Closed)
                window.close();
        }

        window.clear();
        window.display();
    }
}
```

CMakeLists.txt:
```CMake
cmake_minimum_required(VERSION 3.15)

project(Project VERSION 1.0 LANGUAGES CXX)

set(BUILD_SHARED_LIBS OFF)

add_executable(out main.cpp)

set_property(TARGET out PROPERTY
                MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")

set(SFML_USE_STATIC_STD_LIBS ON CACHE BOOL "Use a static c++ runtime" FORCE)
add_subdirectory(SFML)
target_link_libraries(out PRIVATE sfml-graphics)
```

## Expected behavior

The project should build and the configuration of the sfml targets should show the /MT flag in Visual Studio ( in Release mode).

## Actual behavior

The project doesn't link (with the above error message) and visual studio shows that the /MD flag is selected (in Release mode)

## Possible Fix

I fixed the problem by setting the `MSVC_RUNTIME_LIBRARY` property for all SFML targets to `"MultiThreaded$<$<CONFIG:Debug>:Debug>"` in my CMakeLists.txt.  This change probably needs to be copied to the appropriate files inside the SFML Project.

**Comment by JonnyPtn** [comment]
Can you run cmake with the `--trace` option and share the output from that? You should see it going into [this](https://github.com/SFML/SFML/blob/27a4c83ebc724a978e554e77f5f231b887e0bdd9/CMakeLists.txt#L230) condition.

**Comment by baugid** [comment]
I ran cmake with the ```--trace``` option and appended the whole [output](https://github.com/SFML/SFML/files/5029020/out.txt). Additionally I checked ```CMAKE_CXX_FLAGS_RELEASE``` directly after this [loop](https://github.com/SFML/SFML/blob/27a4c83ebc724a978e554e77f5f231b887e0bdd9/CMakeLists.txt#L231).  The only flags set were ``` /O2 /Ob2 /DNDEBUG ```. So it seems like CMake hasn't set the runtime library flag yet. Then I tried manually adding the ```/MT``` flag to ```CMAKE_CXX_FLAGS_RELEASE``` after the loop, but this didn't seem to affect anything.

**Comment by W4RH4WK** [comment]
I just ran into a very similar issue where the include directory was not attached to the library I am building.

```
.
├── src/
│   └── mylib.cpp
├── vendor/
│   └── SFML/
│       ├── bin/
│       ├── include/
│       ├── lib/
│       ├── share/
│       └── CMakeLists.txt
├── CMakeLists.txt
```

```cmake
# ./CMakeLists.txt
cmake_minimum_required(VERSION 3.16)
project(sfmlfuckery)

add_subdirectory(vendor/SFML)

add_library(mylib src/mylib.cpp)
target_link_libraries(mylib PUBLIC sfml-system)
```

```cmake
# ./vendor/SFML/CMakeLists.txt
set(SFML_DIR ${CMAKE_CURRENT_LIST_DIR}/lib/cmake/SFML)
set(SFML_STATIC_LIBRARIES TRUE)
find_package(SFML REQUIRED COMPONENTS system)
```

```
C:\Users\w4rh4wk\Desktop\sfmlfuckery\src\mylib.cpp(1,10): fatal error C1083: Cannot open include file: 'SFML/System.hpp': No such file or directory
```

![image](https://user-images.githubusercontent.com/328798/108731197-9c867380-752c-11eb-8601-9b19c6b8a7be.png)

[trace.txt](https://github.com/SFML/SFML/files/6023100/trace.txt)

I've used snapshot b762276a27efac365968c8330bdb0b5cca44bf52.

**Comment by W4RH4WK** [comment]
After investigating my issue further, I discovered that imported libraries are only available in local scope (where `find_package` is called). This seems to be the default since `SFMLStaticTargets.cmake` is generated by CMake and the corresponding `add_library` call does not feature the `GLOBAL` flag. Therefore, my problem is actually a misconfiguration on my side and has (probably) nothing to do with OP's issue.

**Comment by sam20908** [comment]
Same issue here:
```cmake
project(test CXX)

file(GLOB SRC_FILES "src/*.cpp")
add_executable(${PROJECT_NAME} WIN32 "${SRC_FILES}")
target_include_directories(${PROJECT_NAME} PRIVATE "include")

set(SFML_STATIC_LIBRARIES TRUE)
find_package(SFML COMPONENTS graphics window system main REQUIRED)
target_link_libraries(${PROJECT_NAME} PUBLIC sfml-graphics sfml-window sfml-system sfml-main)

target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_17)

set_property(TARGET ${PROJECT_NAME} PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
```

**Comment by eXpl0it3r** [comment]
Can you give PR #1834 a try?

**Comment by baugid** [comment]
I checked out the branch bugfix/static_runtime_subdirectory and tried to build my example. Sadly this didn't work with the same problem remaining. All SFML modules still use the dynamic runtime library. eventhough the code path to add the flags is hit. The aforementioned workaround does work though. I assume (I don't know enough about the innerworkings of cmake to be shure) that using targets overwrites `CMAKE_CXX_FLAGS_BUILDTYPE` at the end of the execution of cmake.
If you output the variable directly at the end of the code path, the added linker flags are there.
This behaviour coincides with my findings in [this](https://github.com/SFML/SFML/issues/1679#issuecomment-669225056) comment.
Why is it not possible to add the workaround to the main CMake script of SFML?

**Comment by eXpl0it3r** [comment]
Odd, the branch works correctly for me. Maybe you didn't fully clear the cache?
But as the discussion indicates on the PR #1834, I'll probably add the `set_property(... MSVC_RUNTIME_LIBRARY ...)` to SFML directly with some version check. Since it only applies to Windows, it shouldn't be that big of an issue, but raising the minimum version of CMake is a bit tricky, if you want to retain compatibility with older/locked down Linux distros.

**Comment by baugid** [comment]
I deleted the entire build directory so that isn't the difference between our setups.
I just rerun cmake but changed the minimum required version in my cmake script to 3.14 from 3.15 (and fixed the then broken config on my side). Now the build passes.
I then tried to build the master branch with cmake 3.14 and the build passed. 
It seems like some change in cmake 3.15 completely nullifies any changes done directly to the buildflags. 
So this issue only occurs, when someone tries to build SFML with a required cmake version in their scripts of 3.15 or higher.
